task_manager/task_manager/task_client.py

```diff
@@ -185,17 +191,68 @@ def _request_canceling(self, timeout: float) -> List[GoalInfo]:
             )
             raise CancelTaskFailedError("Cancel request timed out.") from e
-        return future.result().goals_canceling
+        return future.result()
+
+    def _handle_cancel_response(self, response: CancelGoal.Response) -> None:
+        """Handles the response from the cancel request.
+
+        :raises UnknownIdAtCancelTaskError: If the goal with known id does not exist
+        :raises GoalAlreadyTerminatedAtCancelTaskError: If the goal has already terminated
+        :raises CancelTaskFailedError: If the cancel request fails
+        """
+        # There seems to be a bug in rclpy, making the return code to be 0 (ERROR_NONE),
+        # no matter if the cancel was rejected or accepted. So checking instead if the
+        # goal is within the cancelling goals.
+        if response.return_code == CancelGoal.Response.ERROR_UNKNOWN_GOAL_ID:
+            self._node.get_logger().info(
+                f"Action server {self.task_specs.topic} did not recognize the goal id. "
+                f"Maybe server has restarted during the task execution and the goal no longer exists. "
+                f"Considering the task canceled."
+            )
+            if not self._result_future:
+                raise CancelTaskFailedError("Couldn't cancel the task, action result future does not exist!")
+            self._result_future.cancel()
+
+        elif response.return_code == CancelGoal.Response.ERROR_GOAL_TERMINATED:
+            self._node.get_logger().info(
+                f"Action server {self.task_specs.topic} did not accept to cancel the goal. "
+                f"Goal seems to have already finished. Considering the task canceled."
+            )
+            if not self._result_future:
+                raise CancelTaskFailedError("Couldn't cancel the task, action result future does not exist!")
+            self._result_future.cancel()
+
+        else:
+            if not self._goal_handle:
+                raise CancelTaskFailedError("Couldn't cancel the task, goal handle does not exist!")
+            goal_ids_cancelling = [goal_info.goal_id for goal_info in response.goals_canceling]
+            if self._goal_handle.goal_id not in goal_ids_cancelling:
+                self._node.get_logger().error(
+                    f"Couldn't cancel the task. Action server {self.task_specs.topic} did not "
+                    f"accept to cancel the goal."
+                )
+                raise CancelTaskFailedError("Couldn't cancel the task!")
 
     def _goal_done_cb(self, future: Future) -> None:
         """Called when the Action Client's goal finishes. Updates the task status and notifies callbacks further the
         other callbacks.
 
         :param future: Future object giving the result of the action call.
         """
+        if future.cancelled():
+            self.task_details.status = TaskStatus.CANCELED
```

Comment:
This is something I went back and forth during the development. I ended up with CANCELED as I believe it is actually requested by the user, sometimes. The way we get here should be via `cancel_task` which can be called straight by user or via proxy for example in case of starting new blocking task. 

Funny enough the case of "something went wrong" is the reason I did not include the check for `cancel_reported_as_success` for this case.

Now this is actually quite tough question... Trying to think that what is the result that the user will see in the end. In the case of `ERROR_GOAL_TERMINATED` most likely there was a small hickup in the communication which caused task manager not to receive the response. In this case `canceled` state makes sense in my mind. On the other hand, the case of `ERROR_UNKNOWN_GOAL_ID` something has really went wrong and it actually would make sense to report `error` to let the user know of the situation. :thinking:

Reply:
Hmm, yeah this makes sense, I somehow forgot for a moment that we actually did end up here via the cancel request. I agree that reporting error in the case of `ERROR_UNKNOWN_GOAL_ID` would be good, so you could add that if you want to. Otherwise, I suppose we can also leave it as is.

Reply:
To separate those errors I'll now have to either call `set_result` or `set_exception`(?) for the future to be able to differentiate the cases in the `goal_done_cb`. :thinking:

Reply:
True, or save the error code in some other way. Let's then maybe just leave it like it is for now, it is not clear to me how meaningful the difference even is. If I remember correctly, we may get the `ERROR_UNKNOWN_GOAL_ID` even in the same case than the `ERROR_GOAL_TERMINATED` if it has been long enough since the goal finished (though I could misremember or it might have changed since I looked at it). So, I don't see reporting it as canceled to be that bad.